staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go

```diff
@@ -0,0 +1,109 @@
+package storage
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/blang/semver/v4"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/component-base/version"
+)
+
+type ResourceSpec struct {
+	// Version indicates the Kubernetes version from which this GVR is decodable
+	K8sVersion semver.Version
+
+	// Version indicates the resource version for which support was added
+	Version string
+}
+
+type ResourceHistory = map[schema.GroupResource][]ResourceSpec
+
+type CompatibilityResourceEncodingConfig struct {
+	binaryVersion        semver.Version
+	compatibilityVersion semver.Version
+	versionSkew          uint8
+	resourceHistory      ResourceHistory
+}
```

Comment:
the types could be defined here, but specific k8s versions and history of served types belongs somewhere like pkg/kubeapiserver, etc, not in the generic k8s.io/apiserver library

Reply:
Moved all relevant code to pkg/kubeapiserver